### PR TITLE
Draft: quality control study cards + ShUp auto-import dicom zip test

### PR DESCRIPTION
**Pull request by michaelkain**
- Extension on study cards: to become study wide quality cards
- Study card conditions: based on dicom tags or database/entity fields + multi-values possible
- Frontend: 
    - 1) modification for multi-value condition (works as before with one value) 
    - 2) tag subject with VALID, WARNING or ERROR in case of quality control study card result
- Minor: OHIF-viewer can be called from study-tree in study detail on exam level
- ShUp: test added to auto-import dicom zip (phantom) multiple times to locally populate the database
- Fix on null pointer exception, when calling pacs for /metadata on re-application for study card
- Update in docker-compose.yml to run on Fedora Linux (ulimits + new version of rabbitmq)
- RabbitMQConfiguration: clean up with queues from hybrid version (shanoir old + shanoir-ng in parallel)

### CODE REVIEW

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/controler/StudyCardApi.java [hunk 1]
```diff
@@ -160,4 +161,17 @@ ResponseEntity<Void> updateStudyCard(
...
+	@ApiOperation(value = "", notes = "Apply a study card on a study for quality control", response = Void.class, tags = {})
+	@ApiResponses(value = {
+		@ApiResponse(code = 200, message = "applied a study card on its study for quality control", response = Void.class),
+		@ApiResponse(code = 401, message = "unauthorized", response = Void.class),
+		@ApiResponse(code = 403, message = "forbidden", response = Void.class),
+		@ApiResponse(code = 422, message = "bad parameters", response = Void.class),
+		@ApiResponse(code = 500, message = "unexpected error", response = Void.class)
+	})
+	@RequestMapping(value = "/apply_on_study/{studyCardId}", method = RequestMethod.GET)
+	@PreAuthorize("hasRole('ADMIN') or (hasRole('EXPERT') and @datasetSecurityService.hasRightOnStudyCard(#studyCardId, 'CAN_ADMINISTRATE'))")
```
Comment (jcomedouteau) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/controler/StudyCardApi.java:173]: Only study administrator can apply a quality control ? (seems legit actually)

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java [hunk 2]
```diff
@@ -14,108 +14,355 @@
...
+	public void applyStudyCard(StudyCardApply studyCardApplyObject) {
+		StudyCard studyCard = studyCardService.findById(studyCardApplyObject.getStudyCardId());
+		LOG.debug("re-apply studycard n° " + studyCard.getId());
+		List<DatasetAcquisition> acquisitions = datasetAcquisitionService.findById(studyCardApplyObject.getDatasetAcquisitionIds());
+		boolean changeInAtLeastOneAcquisition = false;
+		for (DatasetAcquisition acquisition : acquisitions) {
+			if (CollectionUtils.isNotEmpty(acquisition.getDatasets())) {
+				Dataset refDataset = acquisition.getDatasets().get(0);
+				Attributes dicomAttributes = getDicomAttributesForDataset(refDataset);
+				if (studyCard.getRules() != null) {
```
Comment (jcomedouteau) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java:124]: Test for studyCard rules emptyness before loading all dicom attributes ?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java [hunk 3]
```diff
@@ -14,108 +14,355 @@
...
+			boolean conditionVerifiedOnAtLeastOneAcquisition = false;
+			for (DatasetAcquisition acquisition: acquisitions) {
+				int getDicomTagOrField = condition.getDicomTagOrField();
+				// A) check for a dicom tag using a metadata call to the pacs
+				if (Field.getEnum(getDicomTagOrField) == null) {
+					if (CollectionUtils.isNotEmpty(acquisition.getDatasets())) {
+						Dataset refDataset = acquisition.getDatasets().get(0);
+						Attributes dicomAttributes = getDicomAttributesForDataset(refDataset);
+						if (dicomConditionFulfilled(condition, dicomAttributes)) {
+							conditionVerifiedOnAtLeastOneAcquisition = true;
```
Comment (jcomedouteau) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java:219]: Break the for loop at this moment ?

Why do we check only on "at least on acquisition" and not all ?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java [hunk 4]
```diff
@@ -14,108 +14,355 @@
...
+					return true; // as condition values are combined by OR: return if one is true
+				}
+			} else if (tagType.isTextual()) {
+				if (!condition.getOperation().isTextual()) {
+					throw new IllegalArgumentException("Study card processing : operation " + condition.getOperation() + " is not compatible with dicom tag " 
+							+ condition.getDicomTagOrField() + " of type " + tagType + "(condition id : " + condition.getId() + ")");
+				}	
+				String stringValue = dicomAttributes.getString(condition.getDicomTagOrField());
+				if (stringValue == null) {
+					LOG.warn("Could not find a value in the dicom for the tag " + condition.getDicomTagOrField());
```
Comment (jcomedouteau) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java:355]: Do we have a way to test emptyness of a tag ?

File: shanoir-ng-studies/src/main/java/org/shanoir/ng/subjectstudy/model/SubjectStudyStudyCardTag.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		return type;
+	}
+
+	public void setType(int type) {
+		this.type = type;
+	}
+	
+	public Tag getTag() {
+		Tag tag = new Tag();
+		if (type == 1) {
```
Comment (jcomedouteau) [inline shanoir-ng-studies/src/main/java/org/shanoir/ng/subjectstudy/model/SubjectStudyStudyCardTag.java:40]: Add some constants ? (and use them further in code ?)

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/configuration/ShanoirDatasetsConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.scheduling.annotation.EnableAsync;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskExecutor;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+@Configuration
+@EnableAsync
+public class ShanoirDatasetsConfiguration {
```
Comment (jcomedouteau) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/configuration/ShanoirDatasetsConfiguration.java:13]: What is it used for ?

Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/configuration/ShanoirDatasetsConfiguration.java]: Hi Jean-Come, I removed it. I think it was an experiment, when I thought about Spring Batch...

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/controler/StudyCardApi.java (hunk 1 above)
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/controler/StudyCardApi.java:174]: Yes, I would say so.

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java (hunk 2 above)
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java:126]: Yes, well seen, done.

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java (hunk 3 above)
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java:231]: Yes, done. Well seen!

File: shanoir-ng-studies/src/main/java/org/shanoir/ng/subjectstudy/model/SubjectStudyStudyCardTag.java (hunk 5 above)
Comment (michaelkain) [reply shanoir-ng-studies/src/main/java/org/shanoir/ng/subjectstudy/model/SubjectStudyStudyCardTag.java:40]: not sure, what a constant might change here as there is no other usage in code and the entire method is only a work-around to change the GUI display according to the type, to have a display beta version without spending to much time on it

**Comment by julien-louis** [comment]
Notes

* new type of study card : quality study card
    * bound to a study, not to equipments or centers
    * manage it in a tab in study
        * process the control from here
        * display the result
    * UI to edit the quality study cards

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/StudyCardOnStudyResult.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
+package org.shanoir.ng.studycard.dto;
+
+/**
```
Comment (julien-louis) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/StudyCardOnStudyResult.java:3]: It's missing the license header

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/StudyCardOnStudyResult.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+ * In the idea, that it does not make sense to display an error on a dataset,
+ * when there is already a sequence missing on the exam. Display, that at first
+ * and then display the dataset result, when that problem solved?
+ * 
+ * @author mkain
+ *
+ */
+public class StudyCardOnStudyResult {
+
+	private String subjectName;
```
Comment (julien-louis) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/StudyCardOnStudyResult.java:20]: Remark : If we want to provide links to datasets, acquisitions or examination in the GUI, we would have to add ids here.

File: shanoir-ng-front/src/app/study-cards/shared/study-card.dto.model.ts [hunk 9]
```diff
@@ -57,21 +60,21 @@ export class StudyCardDTO {
...
+    dicomTagOrField: number;
+    operation: Operation;
+    values: StudyCardConditionValueDTO[];
+}
 
-    dicomTag: number;
-	dicomValue: string;
-	operation: Operation;
+export class StudyCardConditionValueDTO {
+    value: string;
```
Comment (julien-louis) [inline shanoir-ng-front/src/app/study-cards/shared/study-card.dto.model.ts:74]: A class for just one string ?

File: shanoir-uploader/src/main/java/org/shanoir/uploader/model/dto/StudyCardOnStudyResultDTO.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
+package org.shanoir.uploader.model.dto;
+
+public class StudyCardOnStudyResultDTO {
```
Comment (julien-louis) [inline shanoir-uploader/src/main/java/org/shanoir/uploader/model/dto/StudyCardOnStudyResultDTO.java:3]: I think it's not a dto. I know that StudyCardOnStudyResult in the datasets ms is in a package named "dto" but I think that package is not well named either.

**Review by julien-louis — CHANGES_REQUESTED** [review]

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/StudyCardConditionValue.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+import javax.persistence.Entity;
+import javax.validation.constraints.NotNull;
+
+import org.hibernate.annotations.GenericGenerator;
+import org.shanoir.ng.shared.core.model.AbstractEntity;
+
+@Entity
+@GenericGenerator(name = "IdOrGenerate", strategy = "org.shanoir.ng.shared.model.UseIdOrGenerate")
+public class StudyCardConditionValue  extends AbstractEntity {
```
[the remark this answers is not in the document]
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/StudyCardConditionValue.java:11]: Dear Julien, please read in the referencing class private List<StudyCardConditionValue> values; is a list of values, not a single value.

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java [hunk 12]
```diff
@@ -14,108 +14,384 @@
...
 		}
 		acquisition.setStudyCard(studyCard);
 		acquisition.setStudyCardTimestamp(studyCard.getLastEditTimestamp());
+	}	
+
+	/**
+	 * Re-application on using web GUI and list of dataset acquisitions.
+	 * @param studyCardApplyObject
+	 */
+	public void applyStudyCard(StudyCardApply studyCardApplyObject) {
```
[the remark this answers is not in the document]
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java:117]: Both methods use a similar code basis and I think it is not the idea of a RESTController to contain so much service logic?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java [hunk 13]
```diff
@@ -14,108 +14,384 @@
...
+		StudyCard studyCard = studyCardService.findById(studyCardApplyObject.getStudyCardId());
+		LOG.debug("re-apply studycard n° " + studyCard.getId());
+		List<DatasetAcquisition> acquisitions = datasetAcquisitionService.findById(studyCardApplyObject.getDatasetAcquisitionIds());
+		boolean changeInAtLeastOneAcquisition = false;
+		for (DatasetAcquisition acquisition : acquisitions) {
+			if (CollectionUtils.isNotEmpty(acquisition.getDatasets())) {
+				if (studyCard.getRules() != null) {
+					Dataset refDataset = acquisition.getDatasets().get(0);
+					Attributes dicomAttributes = getDicomAttributesForDataset(refDataset);
```
[the remark this answers is not in the document]
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/service/StudyCardProcessingService.java:126]: no, it is not possible

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/SubjectStudyStudyCardTag.java [hunk 14]
```diff
@@ -0,0 +1,25 @@
+package org.shanoir.ng.studycard.dto;
+
+public class SubjectStudyStudyCardTag {
```
[the remark this answers is not in the document]
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/SubjectStudyStudyCardTag.java:3]: This class is only a memory object, nothing in the database and the idea is to link the Long ID with a type, as small as possible to know what to display in the GUI

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/StudyCardCondition.java [hunk 15]
```diff
@@ -14,43 +14,40 @@
...
+	private int dicomTagOrField;
 	
-	private String dicomValue;
+	@OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
+	@JoinColumn(name="condition_id")
+	private List<StudyCardConditionValue> values;
 	
 	@NotNull
 	private int operation;
-
```
[the remark this answers is not in the document]
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/StudyCardCondition.java:38]: hm, yes, I know what you mean. Yes, I am totally sure: field enums will never ever leave between 1 and 99, never, and DICOM tags are (0002,0002), as int value like "1573009", so no risk of mix up. If you see a better option, why not. But I think having each time another value in another column that is empty does not make it much better.

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/StudyCardRule.java [hunk 16]
```diff
@@ -36,9 +36,12 @@ public class StudyCardRule extends AbstractEntity {
 	@JoinColumn(name="rule_id")
 	private List<StudyCardAssignment> assignments;
 	
-@	OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+	@OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
 	@JoinColumn(name="rule_id")
 	private List<StudyCardCondition> conditions;
+
+	@NotNull
+	private int type; // examination, acquisition, dataset
```
[the remark this answers is not in the document]
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/StudyCardRule.java:44]: Yes, at the moment it is like that. Each rule has its level == type on which it applies, with all its conditions. For me a new rule has to be added on different levels, not only a condition. Each condition validates either a DICOM value in the PACS or a field in the mysql database of shanoir. See, the GUI proposal of Jean-Come I understand it like that.

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/StudyCardOnStudyResult.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
+package org.shanoir.ng.studycard.dto;
+
+/**
+ * This class contains the result of an application of a study card
+ * on an entire study. For each examination, when the result is wrong
```
[the remark this answers is not in the document]
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/StudyCardOnStudyResult.java:5]: for me it belongs here

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/StudyCardOnStudyResult.java (hunk 7 above)
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/StudyCardOnStudyResult.java:3]: I fully agree, should I add it manually or will we add it with your tool?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/StudyCardOnStudyResult.java (hunk 8 above)
Comment (michaelkain) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/dto/StudyCardOnStudyResult.java:20]: I agree, I will do so

File: shanoir-ng-front/src/app/study-cards/shared/study-card.dto.model.ts (hunk 9 above)
Comment (michaelkain) [reply shanoir-ng-front/src/app/study-cards/shared/study-card.dto.model.ts:74]: see above, list of values

File: shanoir-uploader/src/main/java/org/shanoir/uploader/model/dto/StudyCardOnStudyResultDTO.java (hunk 10 above)
Comment (michaelkain) [reply shanoir-uploader/src/main/java/org/shanoir/uploader/model/dto/StudyCardOnStudyResultDTO.java:3]: hm, for me it is a DTO that only purpose is to be transferred to the GUI to be displayed?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/StudyCardConditionValue.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (julien-louis) [reply shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/StudyCardConditionValue.java:11]: Can several conditions share the same value id ?

File: shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/StudyCardCondition.java [hunk 18]
```diff
@@ -14,43 +14,40 @@
...
+	private static final long serialVersionUID = 6708177853555591193L;
 	
-	@NotNull
-	private int dicomTag;
+	private int dicomTagOrField;
 	
-	private String dicomValue;
+	@OneToMany(fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true)
+	@JoinColumn(name="condition_id")
+	private List<StudyCardConditionValue> values;
```
Comment (julien-louis) [inline shanoir-ng-datasets/src/main/java/org/shanoir/ng/studycard/model/StudyCardCondition.java:40]: I think you can use @ElementCollection instead

```
@ElementCollection 
private List<String> values;
```

should suffice, from what I see elsewhere in the code (ie MrProtocolMetadata)

**Comment by michaelkain** [comment]
Now all on Julien's PR.

**Pull request closed without merging**